// parse-comment-es6/dist/stripComment.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = stripComment;
function stripComment(stringIN) {
    var option = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    var SLASH = '/';
    var BACK_SLASH = '\\';
    var STAR = '*';
    var DOUBLE_QUOTE = '"';
    var SINGLE_QUOTE = "'";
    var NEW_LINE = '\n';
    var CARRIAGE_RETURN = '\r';
    var BACK_QUOTE = '`';

    var string = stringIN;
    var length = string.length;
    var position = 0;
    var output = [];

    var comments = [];
    var lineNumber = 0;
    var lineStart = 0;

    var comment = option.comment === true;
    var range = option.range === true;
    var loc = option.loc === true;
    var raw = option.raw === true;

    function nextLine() {
        lineNumber += 1;
    }

    function getCurrentCharacter() {
        return string.charAt(position);
    }

    function getPreviousCharacter() {
        return string.charAt(position - 1);
    }

    function getNextCharacter() {
        return string.charAt(position + 1);
    }

    function add() {
        output.push(getCurrentCharacter());
    }

    function next() {
        position += 1;
    }

    function atEnd() {
        return position >= length;
    }

    function addComment(com) {
        if (!comment) {
            return;
        }

        if (!raw) {
            delete com.raw; // eslint-disable-line
        }

        if (!loc) {
            delete com.loc; // eslint-disable-line
        }

        if (!range) {
            delete com.range; // eslint-disable-line
        }
        comments.push(com);
    }

    function isEscaping() {
        if (getPreviousCharacter() === BACK_SLASH) {
            var caret = position - 1;
            var escaped = true;
            while (caret - 1 >= 0) {
                caret -= 1;
                if (string.charAt(caret) !== BACK_SLASH) {
                    return escaped;
                }
                escaped = !escaped;
            }
            return escaped;
        }
        return false;
    }

    function processSingleLineComment() {
        if (getCurrentCharacter() === SLASH) {
            if (getNextCharacter() === SLASH) {
                var com = {
                    type: 'LineComment',
                    loc: {},
                    range: {}
                };
                com.loc.start = {
                    line: lineNumber,
                    column: position - lineStart
                };
                com.range.start = position;
                next();
                while (!atEnd()) {
                    next();
                    if (getCurrentCharacter() === NEW_LINE || getCurrentCharacter() === CARRIAGE_RETURN) {
                        com.loc.end = {
                            line: lineNumber,
                            column: position - lineStart
                        };
                        com.range.end = position;
                        com.raw = string.substring(com.range.start, com.range.end);
                        nextLine();
                        if (getCurrentCharacter() === CARRIAGE_RETURN && getNextCharacter() === NEW_LINE) {
                            add();
                            next();
                        }
                        add();
                        next();
                        addComment(com);
                        lineStart = position;
                        return true;
                    }
                }
                com.loc.end = {
                    line: lineNumber,
                    column: position - lineStart
                };
                com.range.end = position;
                com.raw = string.substring(com.range.start, com.range.end);
                addComment(com);
            }
        }
        return false;
    }

    function processMultiLineComment() {
        if (getCurrentCharacter() === SLASH) {
            if (getNextCharacter() === STAR) {
                var com = {
                    type: 'BlockComment',
                    loc: {},
                    range: {}
                };
                com.loc.start = {
                    line: lineNumber,
                    column: position - lineStart
                };
                com.range.start = position;
                next();
                while (!atEnd()) {
                    next();
                    if (getCurrentCharacter() === NEW_LINE || getCurrentCharacter() === CARRIAGE_RETURN) {
                        nextLine();
                        if (getCurrentCharacter() === CARRIAGE_RETURN && getNextCharacter() === NEW_LINE) {
                            next();
                        }
                        lineStart = position + 1;
                    } else if (getCurrentCharacter() === STAR) {
                        if (getNextCharacter() === SLASH) {
                            next();
                            next();
                            com.loc.end = {
                                line: lineNumber,
                                column: position - lineStart
                            };
                            com.range.end = position;
                            com.raw = string.substring(com.range.start, com.range.end);
                            addComment(com);
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    function processSingleCharacterExpression(character) {
        if (getCurrentCharacter() === character) {
            add();
            next();
            while (!atEnd()) {
                /**
                 * tolerate line feed
                 */
                if (getCurrentCharacter() === NEW_LINE || getCurrentCharacter() === CARRIAGE_RETURN) {
                    nextLine();
                    if (getCurrentCharacter() === CARRIAGE_RETURN && getNextCharacter() === NEW_LINE) {
                        add();
                        next();
                    }
                    add();
                    next();
                    lineStart = position;
                    if (character !== BACK_QUOTE) {
                        return true;
                    }
                } else if (getCurrentCharacter() === character && !isEscaping()) {
                    add();
                    next();
                    return true;
                } else {
                    add();
                    next();
                }
            }
        }
        return false;
    }

    function process() {
        if (processSingleCharacterExpression(BACK_QUOTE)) {
            return;
        }
        if (processSingleCharacterExpression(DOUBLE_QUOTE)) {
            return;
        }
        if (processSingleCharacterExpression(SINGLE_QUOTE)) {
            return;
        }
        if (processSingleLineComment()) {
            return;
        }
        if (processMultiLineComment()) {
            return;
        }
        if (processSingleCharacterExpression(SLASH)) {
            return;
        }
        if (!atEnd()) {
            if (getCurrentCharacter() === NEW_LINE || getCurrentCharacter() === CARRIAGE_RETURN) {
                nextLine();
                if (getCurrentCharacter() === CARRIAGE_RETURN && getNextCharacter() === NEW_LINE) {
                    add();
                    next();
                }
                add();
                next();
                lineStart = position;
            } else {
                add();
                next();
            }
        }
    }

    while (!atEnd()) {
        process();
    }
    return {
        text: output.join(''),
        comments: comments
    };
}
